### DISCUSSION TITLE
Remove block comments (to help lexing)

### DISCUSSION THREAD

**Discussion by JonathanStarup** [flix/flix#6531] [Ideas]
Block comments make lexing complex. what happens here for example
```flix
/*
/*
def example(): String = "/* */"
*/
*/
```

See this issue for more discussion https://github.com/flix/flix/issues/4947

We could:
- Remove block comments
- Do my crazy asterisk nesting idea https://github.com/flix/flix/issues/4947
  - The idea is: no nested blocks, but `/*` is only closed by `*/`, and this is also true for `/**` and `**/` and so on. So you can nest block comments with different amounts of asteriscs.

**Comment by magnus-madsen** [comment]
Strangely, I would support, supporting up to two levels of block comments. But I suppose that requires the same infrastructure.

**Reply by JonathanStarup** [reply]
Yeah. unless you mark it with `/** **/`

**Reply by magnus-madsen** [reply]
Does your proposal require a lot of code?

**Reply by JonathanStarup** [reply]
1) They do not have to be properly "numbered". This is a valid block comment `/* /*** ***/ **/ **/ */`
2) The code should be simple: when the lexer sees an opening block comment (a `/` with some positive number, k, of `*`), then just scan until you find k `*` followed by `/`.

This means that the block comment part only needs k look-ahead, which is technically unbounded but very simple in practice

**Comment by magnus-madsen** [comment]
I think I am coming around to Jonathans proposal _if_ it simplifies the implementation.

**Comment by mlutze** [comment]
Have we established why we need block comments? Any IDE can easily comment-out a block with line comments.

**Reply by JonathanStarup** [reply]
Yeah, I can only think of non-ide people. It also feels like flix doesn't like block comments by the fact that doc comments are `///`

**Comment by stephentetley** [comment]
I quite like block comments as I'm often having to comment out chunks of code and blocks seem easier to keep track off rather than nested line comments.

That said, if they make parsing and lexing significantly harder I'd be okay without them.

**Reply by jaschdoc** [reply]
One could argue that not supporting block comments encourages developers to use their version control system more... I'm not sure I'm convinced that's true, though.

**Reply by magnus-madsen** [reply]
How so?

**Reply by jaschdoc** [reply]
If it's impossible to use block comments, then it may be easier to commit something that works, delete the stuff that you would normally comment out and experiment with that change, then commit again when it works (or stashing / restoring if it doesn't work). It's just a hypothesis, though. I'm not convinced it's entirely true hehe :)

**Comment by magnus-madsen** [comment]
According to @herluf-ba, the complexity of block comments is not too bad. It is mostly string interpolation that is difficult/more complex.

The code to handle block comments is really only ~50 lines. We have lots of stuff that is more complex than that.

So at least for now, I think we should keep them.

Do note that for "documentation comments" one should use `///`.

**Reply by JonathanStarup** [reply]
Sorry to just ask instead of checking, but the lexer doesn't handle these cases, I assume?
```
/* def example(): String = "*//" */
/* def example(): String = "//*" */
```

**Reply by herluf-ba** [reply]
Thanks for these cases, It does not handle those right now!
First one gives
```
-- Parse Error -------------------------------------------------- foo.flix

>> Parse Error: Invalid input "/"", expected Use, Import, "mod", PureWS, PossibleDocComment, Annotations, Modifiers or "def" (line 1, column 31):
/* def example(): String = "*//" */
                              ^

Syntactic Context: OtherDecl.

-- Lexer Error -------------------------------------------------- foo.flix

>> missing '"' in string.

1 | /* def example(): String = "*//" */
                                   ^
                                   String starts here.
```
So this one is already a `ParseError` according to the current parser.
Ideally it should just be a block-comment in my opinion.

The second one gives:
```
-- Lexer Error -------------------------------------------------- foo.flix

>> Missing '*/' in block-comment.

1 | /* def example(): String = "//*" */
    ^
    Block-comment starts here.
```
This should just be a block-comment.

This issue stems from the fact that within a block-comment it does matter which context a `/*` or `*/` appears in. As of now `/*` and `*/` are only allowed within a string, so the block-comment tokenizer will also need to know if it is within a string.

**Reply by herluf-ba** [reply]
To be clear: I am working on handling these cases in the lexer. There will be a PR soon.

**Reply by herluf-ba** [reply]
Not really complicated to handle: https://github.com/flix/flix/pull/6553

**Reply by JonathanStarup** [reply]
Due to Matt: What about this program
```
/*   OBS: we need to handle ` " `   */
```
Now this is a lexer error (probably at least, or some variant of it will be).

---

I think the nested block comments problem is fundamentally unsolvable because you cannot distinguish commented-out code and comments that happen to look like code.

**Comment by athas** [comment]
Clearly the solution is to not allow nesting, *but* to allow the user to specify their own end-of-comment symbol, similar to shell [heredocs](https://en.wikipedia.org/wiki/Here_document#Unix_shells).  This symbol can then be picked as a GUID (or perhaps a cryptographic hash of the commented-out code?), which is essentially guaranteed not to be already present in the code.  This is an elegant and perfectly robust solution for people who really don't want to use line comments.

**Comment by athas** [comment]
Another observation: Flix already has a serviceable facility for putting things in source code that will be ignored in the compiler (line comments).  Yet apparently there is still enough need and interest for another facility, doing the same thing, to sustain several issues, pull requests, and design discussions, none of which have been deemed acceptable so far.  It would be very demotivating if all this activity did not result in an advanced language or compiler feature, would it not?  To address the problem of what happens when commented-out code contains an end-of-comment marker, why not introduce nondeterminism in the lexer/parser?  Whenever the parser encounters an block comment end/start marker, it first tries to ignore it and see if that results in a valid parse, then respects it and checks if *that* results in a valid parse.  (If both succeed, that should probably be an error.)  As an optimisation, you could also look a few characters back and forth to check whether it's likely you're inside a string literal.  Maybe an LLM could be trained to make this guess more accurate?

Another elegant solution is to require all comments to be lexically valid Flix code.  Then you can easily recognise string literals and ignore any embedded commenting directives.  It also follows modern trends in language implementations being more strict about things that used to be accepted (e.g. not having warnings, only errors).  I think this approach certainly merits further consideration.  Perhaps a Flix Language Design Subcommittee should be established to study block comments more carefully?  One of the most important productivity properties of a language is after all how good its facilities are for making the compiler ignore parts of its input.

**Reply by magnus-madsen** [reply]
> Another elegant solution is to require all comments to be lexically valid Flix code... [...]

Haha. I am dying. I can't breathe.

**Comment by magnus-madsen** [comment]
Another argument against block comments: https://matklad.github.io/2023/08/01/on-modularity-of-lexical-analysis.html

**Reply by bungcip** [reply]
beside zig, i think moonbit lang also don't have token that span multiple line (meaning its don't have block comment and  using `#| ` for multi line string (https://docs.moonbitlang.com/en/latest/language/fundamentals.html#string))

**Comment by JonathanStarup** [comment]
@athas here's more ammunition from rust [#129685](https://github.com/rust-lang/rust/issues/129685)

**Comment by athas** [comment]
For this particular battle, I think my quiver cannot fit any more arrows.
